src/algorithms/contractions/ctmrg_contractions.jl

```diff
@@ -776,17 +770,29 @@ Apply projectors to each side of a quadrant.
         |
 ```
 """
-function renormalize_corner(
-    quadrant::AbstractTensorMap{T,S,3,3}, P_left, P_right
-) where {T,S}
-    return @autoopt @tensor corner[χ_in; χ_out] :=
-        P_right[χ_in; χ1 D1 D2] * quadrant[χ1 D1 D2; χ2 D3 D4] * P_left[χ2 D3 D4; χ_out]
-end
-function renormalize_corner(
-    quadrant::AbstractTensorMap{T,S,2,2}, P_left, P_right
-) where {T,S}
-    return @autoopt @tensor corner[χ_in; χ_out] :=
-        P_right[χ_in; χ1 D1] * quadrant[χ1 D1; χ2 D3] * P_left[χ2 D3; χ_out]
+@generated function renormalize_corner(
+    quadrant::AbstractTensorMap{T,S,N,N}, P_left, P_right
```

Comment:
Can we be more restricted with the generated functions: 
```suggestion
    quadrant::AbstractTensorMap{<:Any,S,N,N}, P_left::AbstractTensorMap{<:Any,S,N,1}, P_right::AbstractTensorMap{<:Any,S,1,N}
```

or at least make it `AbstractTensorMap`?